File: lib/python3/PUJ/Model/Logistic.py
## =========================================================================
## @author Leonardo Florez-Valencia (florez-l@javeriana.edu.co)
## =========================================================================

import numpy
from .Linear import *

'''
'''
class Logistic( Linear ):

  ## -----------------------------------------------------------------------
  ## Initialize an object witha zero-sized parameters vector
  ## -----------------------------------------------------------------------
  def __init__( self ):
    pass
  # end if

  ## -----------------------------------------------------------------------
  ## Final-user methods.
  ## -----------------------------------------------------------------------
  def evaluate( self, x ):
    return 1.0 / ( 1.0 + numpy.exp( -super( ).evaluate( x ) ) )
  # end def

  def threshold( self, x ):
    z = self.evaluate( x )
    return ( z >= 0.5 ).astype( z.dtype )
  # end def

  ## -----------------------------------------------------------------------
  '''
  Cross cost function for linear regressions.
  '''
  class Cost( Base.Cost ):

    ## ---------------------------------------------------------------------
    ## Initialize an object witha zero-sized parameters vector
    ## ---------------------------------------------------------------------
    def __init__( self, model, X, y ):
      super( ).__init__( model, X, y )
      self.m_mY = y.mean( )
      self.m_XhY = numpy.matrix( numpy.multiply( X, y ).mean( axis = 0 ) ).T
    # end def

    ## ---------------------------------------------------------------------
    ## Evaluate cost with gradient (if needed)
    ## ---------------------------------------------------------------------
    def evaluate( self, need_gradient = False ):
      z = self.m_Model.evaluate( self.m_X )
      J  = numpy.log( z[ self.m_Y == 1 ] + 1e-12 ).sum( )
      J += numpy.log( 1 - z[ self.m_Y == 0 ] + 1e-12 ).sum( )
      J /= -float( self.m_X.shape[ 0 ] )
      if need_gradient:

        g = numpy.zeros( self.m_Model.parameters( ).shape )
        g[ 0 , 0 ] = z.mean( ) - self.m_mY
        g[ 1 : , : ] = \
          numpy.matrix( numpy.multiply( self.m_X, z ) ).mean( axis = 0 ).T - \
          self.m_XhY
        return [ J, g ]
      else:
        return [ J, None ]
      # end if
    # end def
  # end class
# end class

## eof - $RCSfile$

File: lib/cxx/PUJ_ML/Model/Logistic.cxx
// =========================================================================
// @author Leonardo Florez-Valencia (florez-l@javeriana.edu.co)
// =========================================================================

#include <PUJ_ML/Model/Logistic.h>
#include <cmath>

// -------------------------------------------------------------------------
template< class _T >
PUJ_ML::Model::Logistic< _T >::Cost::
Cost( Superclass* model, const TMatrix& X, const TMatrix& Y )
  : Superclass::Cost( model, X, Y )
{
}

// -------------------------------------------------------------------------
template< class _T >
_T PUJ_ML::Model::Logistic< _T >::Cost::
operator()( _T* g ) const
{
  return( _T( 0 ) );
}

// -------------------------------------------------------------------------
template< class _T >
PUJ_ML::Model::Logistic< _T >::
Logistic( )
  : Superclass( )
{
}

// -------------------------------------------------------------------------
template< class _T >
PUJ_ML::Model::Logistic< _T >::
Logistic( const TMatrix& X, const TCol& Y )
  : Superclass( )
{
}

// -------------------------------------------------------------------------
template< class _T >
typename PUJ_ML::Model::Logistic< _T >::
TMatrix PUJ_ML::Model::Logistic< _T >::
operator()( const TMatrix& x )
{
  static const auto f = []( _T z ) -> _T
    {
      static const _T _0 = _T( 0 );
      static const _T _1 = _T( 1 );
      static const _T _b = _T( 40 );

      if     ( z >  _b ) return( _1 );
      else if( z < -_b ) return( _0 );
      else               return( _1 / ( _1 + std::exp( -z ) ) );
    };
  return( this->Superclass::operator()( x ).unaryExpr( f ) );
}

// -------------------------------------------------------------------------
template< class _T >
typename PUJ_ML::Model::Logistic< _T >::
TMatrix PUJ_ML::Model::Logistic< _T >::
operator[]( const TMatrix& x )
{
  return(
    ( this->operator()( x ).array( ) >= _T( 0.5 ) ).template cast< _T >( )
    );
}

// -------------------------------------------------------------------------
template class PUJ_ML_EXPORT PUJ_ML::Model::Logistic< float >;
template class PUJ_ML_EXPORT PUJ_ML::Model::Logistic< double >;
template class PUJ_ML_EXPORT PUJ_ML::Model::Logistic< long double >;

// eof - $RCSfile$
